test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,170 @@
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "badVersion")
+		_, err := framework.Global.KubeClient.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		require.NoError(t, err)
+		if err == nil {
+			break
+		}
+	}
+
+	err = testCtx.waitForWindowsNodes(1, true, false, true)
+	if err != nil {
+		t.Fatal(t, "windows node upgrade failed  with %v", err)
+	}
+	// Get all the Windows nodes
+	nodes, err = framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node missing version annotation")
+			assert.Equal(t, operatorVersion, node.Annotations[nc.VersionAnnotation], "Node recreation failed. WMCO version annotation mismatch")
+		})
+	}
+}
+
+// configureUpgradeTest carries out steps required before running tests for upgrade scenario.
+// The steps include -
+// 1. Scale down the operator to 0.
+// 2. Change Windows node version annotation to an invalid value
+// 3. Scale up the operator to 1
+func configureUpgradeTest() error {
+	// Scale down the WMCO deployment to 0
+	if err := scaleDeployment(0); err != nil {
+		return err
+	}
+
+	// Override the Windows Node Version Annotation
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	if err != nil {
+		return err
+	}
+	if len(nodes.Items) != int(gc.numberOfNodes) {
+		return errors.Wrapf(nil, "unexpected number of nodes %v", gc.numberOfNodes)
+	}
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "badVersion")
+		_, err := framework.Global.KubeClient.CoreV1().Nodes().Patch(context.TODO(), node.Name, types.MergePatchType, []byte(patchData), metav1.PatchOptions{})
+		if err != nil {
+			return err
+		}
+		log.Printf("Node Annotation changed to %v", node.Annotations[nc.VersionAnnotation])
+	}
+
+	// Scale up the WMCO deployment to 1
+	if err := scaleDeployment(1); err != nil {
+		return err
+	}
+	return nil
+}
+
+// scaleDeployment scales the WMCO operator to the given replicas.
+func scaleDeployment(desiredReplicas int32) error {
+	err := wait.Poll(deploymentRetryInterval, deploymentTimeout, func() (done bool, err error) {
+		deployment, err := framework.Global.KubeClient.AppsV1().Deployments(resourceName).Get(context.TODO(), resourceName,
+			metav1.GetOptions{})
+
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Printf("error getting deployment %v", err)
+				return false, nil
+			}
+			return false, err
+		}
+		if deployment.Spec.Replicas != &desiredReplicas {
```

Comment:
Not sure if this check is needed. The patch won't take effect if `deployment.Spec.Replicas == &desiredReplicas`.

Reply:
I added that check because, we won`t need to patch the WMCO deployment if the number of replicas is equal to the given input.

Reply:
True, I am just calling out it is a bit redundant. I am fine if you want to retain it.